### ISSUE TITLE
Format resolving and Template lookup

### ISSUE THREAD

**Issue by mereghost** [hanami/hanami#663]
As was reported on Gitter, requests that are made with a `Content-Type` of `text/plain` get resolved to the `:conf` format and consequently a `.txt` template is never used.

That happens because we use `Rack::Mime::MIME_TYPES.key(mime_type)` [(see here)](https://github.com/hanami/controller/blob/master/lib/hanami/action/mime.rb#L486) which returns only the first match. The same problem can happen for a bunch of other types (around 40) and we never detected this problem because we skipped those cases on purpose on the tests [(see here)](https://github.com/hanami/controller/blob/master/test/action/format_test.rb#L85)

Making it work is kinda of a huge change. Why? For that to work we'd need to make `Hanami::Action#format` an `Array` or make `Hanami::View` (specially the template finder) MIME aware and make it look for the first template file it finds with the possible extensions.

That would warrant changes in `view`, `controller` and possibly `hanami`, and that brings me here. I'm willing to work at it but I'd like some input from the hanami members on the best way to tackle this.

**Comment by jodosha** [comment]
@mereghost I think the problem is in the lookup from the mime type textual representation (`"text/plain"`) to the symbol (`:txt`). IMO `view` has nothing to do with it, as it only accepts the result of that lookup. So `view` is decoupled from `controller`.

Can you please have a look at `controller`? Thanks!

**Comment by mereghost** [comment]
@jodosha In fact, `controller` is the one responsible for the [behaviour](https://github.com/hanami/controller/blob/master/lib/hanami/action/mime.rb#L486). But a change in controller can lead to a change in template lookup, i.e. if Action pass an array of possible file extensions to `view` so it can look for all possible templates.

A nice way to expose the problem is commenting [this line](https://github.com/hanami/controller/blob/master/test/action/format_test.rb#L85) on `controller` test suite.

Question is: do we pass down a number of possible formats? That would keep `controller` oblivious to `view` code, but will start a bunch of changes namely: `format` would return an array, `view` would need to cope with that, `TemplateFinder` might have to look for multiple files, if it finds more than one what should it do?

**Comment by jodosha** [comment]
@mereghost Please make sure that if we're gonna introduce a list of possible MIME types, views unit tests won't increase the complexity of their setup.

I'm talking about the `let(:view)` block in the testing code that we generate.

---

An alternative approach to investigate is to only use a subset of `Rack::Mime::MIME_TYPES`, by manually picking the most common MIME Types and avoiding key duplicates.

Once we have that subset in place, devs can add their own specific MIME types, which are treated as custom entries. Hanami gives priority to custom MIME types and this should solve the problem reported.

It's a hack, but it has the advantage of:

  * Keep `controller` responsible of returning the best choice
  * Keep `view` simple, without any extra lookup
  * Reduce the memory footprint: `Rack::Mime::MIME_TYPES` has ~600 entries, which is too much for the average web project.

WDYT?

**Comment by mereghost** [comment]
That second approach may make more sense. I'll make a sweep on `Rack::Mime::MIME_TYPES` and post a list here of let's say, 10-20 more common types for further discussion.

**Comment by mereghost** [comment]
Okay, I did some work and tried to narrow the mime types to the ones listed below, mainly JS, HTML, Audio, Image and Video. Did I miss something? 

```json
{
   ".manifest":"text/cache-manifest",
   ".atom":"application/atom+xml",
   ".avi":"video/x-msvideo",
   ".bmp":"image/bmp",
   ".bz":"application/x-bzip",
   ".bz2":"application/x-bzip2",
   ".chm":"application/vnd.ms-htmlhelp",
   ".txt":"text/plain",
   ".css":"text/css",
   ".csv":"text/csv",
   ".flv":"video/x-flv",
   ".gif":"image/gif",
   ".gz":"application/x-gzip",
   ".h264":"video/h264",
   ".html":"text/html",
   ".ico":"image/vnd.microsoft.icon",
   ".ics":"text/calendar",
   ".jpg":"image/jpeg",
   ".js":"application/javascript",
   ".json":"application/json",
   ".mp4":"video/mp4",
   ".mov":"video/quicktime",
   ".mp3":"audio/mpeg",
   ".mp4a":"audio/mp4",
   ".mpg":"video/mpeg",
   ".oga":"audio/ogg",
   ".ogg":"application/ogg",
   ".ogv":"video/ogg",
   ".pdf":"application/pdf",
   ".pgp":"application/pgp-encrypted",
   ".png":"image/png",
   ".psd":"image/vnd.adobe.photoshop",
   ".rtf":"application/rtf",
   ".sh":"application/x-sh",
   ".svg":"image/svg+xml",
   ".swf":"application/x-shockwave-flash",
   ".tar":"application/x-tar",
   ".torrent":"application/x-bittorrent",
   ".tsv":"text/tab-separated-values",
   ".uri":"text/uri-list",
   ".vcs":"text/x-vcalendar",
   ".wav":"audio/x-wav",
   ".webm":"video/webm",
   ".wmv":"video/x-ms-wmv",
   ".woff":"application/font-woff",
   ".woff2":"application/font-woff2",
   ".wsdl":"application/wsdl+xml",
   ".xhtml":"application/xhtml+xml",
   ".xml":"application/xml",
   ".xslt":"application/xslt+xml",
   ".yml":"text/yaml",
   ".zip":"application/zip"
}
```

**Comment by jodosha** [comment]
@mereghost Good to go 👍

**Comment by mereghost** [comment]
Nice! I'll code it in the next few days and send the PR for `controller`

**Comment by jodosha** [comment]
The fix for this issue [introduces a breaking change](https://github.com/hanami/controller/pull/191#discussion_r90884501). We should ship this with v1.0.0.beta1 series, rather than with the 0.9.x one.

**Comment by mereghost** [comment]
@jodosha should we merge this?

**Comment by AlfonsoUceda** [comment]
Closed by hanami/controller#191
